STAR_COUNT_FILE_METAINDEXES = ["N_unmapped", "N_multimapping", "N_noFeature", "N_ambiguous"]
STAR_COUNT_FILE_HEADER = ["Total", "MinusStrand", "PlusStrand"]

# Align and count the fastQ file with STAR
# TODO: make this a thing
def star_align_fastq(srr_id, fastq_file_names, reference_genome, output_path):
    """
    Take a set of FASTQ files and align them with the STAR aligner (in quantMode)

    :param srr_id: str
        NCBI SRR ID string
    :param fastq_file_names: str
        A list of complete FASTQ file names that were unpacked from the SRR file (including path)
    :param reference_genome: str
        A path to the STAR reference genome that was preassembled
    :param output_path: str
        The path to put the output alignment files
    :return count_file_name: str
        The STAR count file generated by --quantMode (including path)
    """
    count_file_name = ""
    return count_file_name
